### PR TITLE
chore: [jsonwebtoken] Add version 7 with difference of SignOptions.expiresIn type

### PR DESCRIPTION
**Pull request by Mifrill**
📓 Based on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68516#pullrequestreview-1901633035.

In this PR created `jsonwebtoken/v7` by copy-paste from `jsonwebtoken/` with a few differences based on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68516, related to `expiresIn` type only.

### CODE REVIEW

File: types/jsonwebtoken/v7/package.json
Comment (Mifrill) [inline types/jsonwebtoken/v7/package.json:4]: 📝 After release from [7.4.3 to 8.0.0](https://github.com/auth0/node-jsonwebtoken/compare/v7.4.3...v8.0.0), SignOptions.expiresIn throw an error when given undefined.

File: types/jsonwebtoken/v7/tsconfig.json
Comment (Mifrill) [inline types/jsonwebtoken/v7/tsconfig.json:13]: 📝 Difference to https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68516/files, in this version we don't need `exactOptionalPropertyTypes` option enabled.

**Comment by jakebailey** [comment]
I have no idea why this PR isn't failing; the tests for the root level `jsonwebtoken` will fail unless its npmignore is updated to _remove_ the subdir that you've added. Probably a bug in the test runner because the parent package won't have been changed.

**Comment by jakebailey** [comment]
I am fixing the tooling to catch this.

**Comment by sandersn** [comment]
the latest version 7 is 7.2.8 -- last published 6 years ago. There were 150k downloads last week, a tiny fraction of the total downloads, but enough that I think that this should get a signoff from at least one code owner since many of those users will now get 7.4.1. How much will this PR cause the code to change from 7.2 -> 7.4? How big is the diff?

**Comment by jakebailey** [comment]
> I am fixing the tooling to catch this.

https://github.com/microsoft/DefinitelyTyped-tools/pull/971

**Comment by Mifrill** [comment]
@sandersn do you mean need to use `index.d.ts` file from version 7.2.8 instead? 
Source: https://www.npmjs.com/package/@types/jsonwebtoken/file/3ee1c4479498ce568cf7789174c181a68a7f7f16f7b338132ded0daf24a8a971

📝 https://www.npmjs.com/package/@types/jsonwebtoken/v/7.2.8?activeTab=code

File: types/jsonwebtoken/v7/index.d.ts
Comment (sandersn) [inline types/jsonwebtoken/v7/index.d.ts:197]: Without any specific knowledge of jsonwebtoken, the addition of these overloads seems like the riskiest part of the change. It's probably OK but I would like somebody who knows jsonwebtoken to take a look too.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
I looked at the v7/index.d.ts and it is completely same as current index.d.ts .. Why do we need separate v7 folder in that case ?

**Comment by jakebailey** [comment]
> I looked at the v7/index.d.ts and it is completely same as current index.d.ts .. Why do we need separate v7 folder in that case ?

The root is v8, v7 is v7 and has different compiler options.

**Comment by jakebailey** [comment]
But, not totally sure I'm understanding why we need to republish v7 when its types were last published quite literally 6 years ago?

**Comment by Mifrill** [comment]
@sheetalkamat regaridng:
> Why do we need separate v7 folder in that case ?

the difference is the compiler option, see: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68891#discussion_r1511836866

@jakebailey 
> But, not totally sure I'm understanding why we need to republish v7 when its types were last published quite literally 6 years ago?

I'm also not sure about that. As I [mentioned earlier](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68516#issuecomment-1935099396):
> it probably makes sense to apply those changes in older versions of a library's type declarations, because that change is affected after release [8.0.0](https://github.com/auth0/node-jsonwebtoken/compare/v7.4.3...v8.0.0).

Could you guys please check [this comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68891#issuecomment-1977558079) and let me know what do you think?

**Comment by jakebailey** [comment]
I'm confused by @sandern's comment; v7 was not the latest at the time of the comment. https://www.npmjs.com/package/@types/jsonwebtoken?activeTab=versions It only has some 60 downloads compared to v8 or v9.

I'd probably just close this PR and let v7 be what it is.

**Comment by sandersn** [comment]
The latest `@types/jsonwebtoken@7` is 7.2.8, which has 150k downloads. This PR updates the 7.* types to 7.4, which means that people who `npm i @types/jsonwebtoken@7` will now get (slightly) different types.

I agree with @jakebailey that updating the types that haven't changed for 6 years isn't valuable for people in general. @Mifrill do you use the 7.* types and need them updated? If not, I don't think there's a need to update them.

**Comment by Mifrill** [comment]
@sandersn Well, yes I'm on `^7.1.9` at this moment, but I don't think it's an issue for me, I just created PR to follow the guide in https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library for changes in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68516, if it's no need, feel free to close it.

**Comment by jakebailey** [comment]
I think that if you're definitely trying to use v7, and want to keep it updated, then Nathan was right and you'd want to copy the index.d.ts and so on from when the package was back on v7, and then go from there.

**Pull request closed without merging**
